File: CourseProjects/CourseProjects/JavaEclipseProjects/DesignPatternsExamples/src/patterns/visitor/HTMLFormatVisitor.java
package patterns.visitor;

import java.io.*;

public class HTMLFormatVisitor implements Visitor {
	private BufferedWriter output;

	public HTMLFormatVisitor(Writer w) {
		output = new BufferedWriter(w);
	}

	public void visitDocumentStart(Document d) {
		write("<html><head/><body>");
	}
	public void visitDocumentEnd() {
		write("</body></html>");
		closeOutput();
	}
	public void visitModuleStart(Module m) {
		write("<h2>New Module</h2>");
	}
	public void visitModuleEnd() {}
	public void visitParagraphStart(Paragraph p) {
		write("<p>" + p + "</p>");
	}
	public void visitParagraphEnd(){}
	public void visitSectionStart(Section s) {
		write("<h3>New Section</h3>");
	}
	public void visitSectionEnd() {}
	public void visitTitleStart(Title title) {
		write("<h4>" + title + "</h4>");
	}
	public void visitTitleEnd(){}
	private void write(String text) {
		try {
			output.write(text);
			output.newLine();
		} catch(IOException ex) {
			System.out.println(ex);
		}
	}
	private void closeOutput() {
		try {
			output.close();
		} catch(IOException ex) {
			System.out.println(ex);
		}
	}
}

File: CourseProjects/CourseProjects/VisualStudioProjects/DesignPatternsExamples/Visitor/HTMLFormatVisitor.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Visitor {
    internal class HTMLFormatVisitor : IVisitor {
        private readonly StreamWriter output;

        public HTMLFormatVisitor(FileStream w) {
            output = new StreamWriter(w);
        }

        public void VisitDocumentStart(Document d) {
            Write("<html><head/><body>");
        }

        public void VisitDocumentEnd() {
            Write("</body></html>");
            CloseOutput();
        }

        public void VisitModuleStart(Module m) {
            Write("<h2>New Module</h2>");
        }

        public void VisitModuleEnd() {}

        public void VisitParagraphStart(Paragraph p) {
            Write("<p>" + p + "</p>");
        }

        public void VisitParagraphEnd() {}

        public void VisitSectionStart(Section s) {
            Write("<h3>New Section</h3>");
        }

        public void VisitSectionEnd() {}

        public void VisitTitleStart(Title title) {
            Write("<h4>" + title + "</h4>");
        }

        public void VisitTitleEnd() {}

        private void Write(String text) {
            output.Write(text);
            output.WriteLine();
        }

        private void CloseOutput() {
            output.Close();
        }
    }
}